packages/app-builder/src/components/hooks/use-introspection/nested-fields.ts

```diff
@@ -0,0 +1,58 @@
+import { IntrospectionOutputTypeRef, IntrospectionObjectType } from 'graphql'
+import { flatKind, getObjectType } from './helpers'
+import { LIST, NON_NULL, OBJECT, QueryableField, SCALAR } from './types'
+
+type NestedDict = {
+  [key: string]: NestedDict | undefined
+}
+
+export const nestedFieldsToString = (nestedFields?: NestedDict) => {
+  if (!nestedFields) {
+    return ''
+  }
+
+  const str = Object.entries(nestedFields)
+    .map(([fieldName, field]) => [fieldName, nestedFieldsToString(field)])
+    .flat()
+    .join(' ')
+
+  return `{ ${str} }`
+}
+
+export const queryableFieldToNestedDict = (
+  type: IntrospectionOutputTypeRef,
+  queryableObjectTypes: IntrospectionObjectType[],
+): NestedDict | undefined => {
+  const nestedDict: NestedDict = {}
+
+  if (type.kind === NON_NULL || type.kind === LIST) {
+    return queryableFieldToNestedDict(type.ofType, queryableObjectTypes)
+  }
+
+  if (type.kind === SCALAR) {
+    return undefined
+  }
+
+  if (type.kind === OBJECT) {
+    const objectType = queryableObjectTypes.find(({ name }) => name === type.name)
+    if (objectType) {
+      objectType.fields.forEach((field) => {
+        nestedDict[field.name] = queryableFieldToNestedDict(field.type, queryableObjectTypes)
+      })
```

Comment:
```suggestion
      return objectType.fields.reduce<NestedDist, string>((dict, field) => ({
       ...dict,
       [field.name]: queryableFieldToNestedDict(field.type, queryableObjectTypes),
      }), nestedDict)
```

Might need to reverse the generic types + parameters - I always get reduce params/types the wrong way around

Reply:
you got it right :D